library/src/main/java/dev/chrisbanes/insetter/Insetter.java

```diff
@@ -332,6 +342,96 @@ public static void setEdgeToEdgeSystemUiFlags(@NonNull final View view, final bo
   }
 
+  /**
+   * Applies padding or margin to the bottom of the view when the IME (keyboard) is displayed
+   * @param v the view to apply inset handling too
+   * @param marginBottomForIme whether to adjust the view's bottom margin
+   * @param paddingBottomForIme whether to adjust the view's bottom padding
+   */
+  public static void applyInsetsWhenImeDisplayed(@NonNull final View v,
+                                                 final boolean marginBottomForIme,
+                                                 final boolean paddingBottomForIme) {
+    setOnApplyInsetsListener(v, new OnApplyInsetsListener() {
+      @Override
+      public void onApplyInsets(@NonNull View view, @NonNull WindowInsetsCompat insets, @NonNull ViewState initialState) {
+        Boolean isIMEOpen = Insetter.isIMEOpen.update(view).get();
+        if (isIMEOpen == null) return;
+
+        int paddingBottom;
+        if (isIMEOpen) {
+          paddingBottom = insets.getSystemWindowInsetBottom();
+        } else {
+          paddingBottom = initialState.getPaddings().getBottom();
+        }
+
+        int marginBottom;
+        if (isIMEOpen) {
+          marginBottom = insets.getSystemWindowInsetBottom();
+        } else {
+          marginBottom = initialState.getMargins().getBottom();
+        }
+
+        if (marginBottomForIme) {
+          final ViewGroup.LayoutParams lp = view.getLayoutParams();
+          if (lp instanceof ViewGroup.MarginLayoutParams) {
+            final ViewGroup.MarginLayoutParams mlp = (ViewGroup.MarginLayoutParams) lp;
+
+            if (mlp.bottomMargin != marginBottom) {
+              mlp.bottomMargin = marginBottom;
+              view.setLayoutParams(lp);
+
+              if (Log.isLoggable(TAG, Log.DEBUG)) {
+                Log.d(
+                        TAG,
+                        String.format(
+                                Locale.US,
+                                "applyInsetsWhenImeDisplayed. Applied margin to %s: bottom=%d}",
+                                view,
+                                marginBottom));
+              }
+            }
+          }
+        }
+
+        if (paddingBottomForIme) {
+          view.setPadding(view.getPaddingLeft(),
+                  view.getPaddingTop(),
+                  view.getPaddingRight(),
+                  paddingBottom);
+
+          if (Log.isLoggable(TAG, Log.DEBUG)) {
+            Log.d(
+                    TAG,
+                    String.format(
+                            Locale.US,
+                            "applyInsetsWhenImeDisplayed. Applied padding to %s: bottom=%d}",
+                            view,
+                            marginBottom));
+          }
+        }
+
+      }
+    });
+  }
+
+  /**
+   * Determines if the IME (keyboard) is open by comparing the visible window height to the height
+   */
+  private static IMEOpen isIMEOpen = new IMEOpen();
+
+  private static class IMEOpen extends ThreadLocal<Boolean> {
```

Comment:
Woah, this isn't what I meant. My bad for now explaining. I meant to cache the rect in a threadlocal so we don't keep creating them. Forget that comment for now and revert back to the previous code :smile: